### DISCUSSION TITLE
[Proposal] Only allow Lexical Contextual Keywords in the Language

### DISCUSSION THREAD

**Discussion by jmarolf** [dotnet/csharplang#4458] [Language Ideas] [answered]
# Only Allow Lexical Keywords in the Language

* [x] Proposed
* [ ] Prototype: 
* [ ] Implementation:
* [ ] Specification:

## Summary
[summary]: #summary

Today there are keywords in the language that cannot be understood with just lexical information such as `var` or `nameof`. These keywords operate this way for backwards compatibility reasons. I propose that we change the language so that there exists no keywords that cannot be determined via lexical analysis. 

## Motivation
[motivation]: #motivation

In general, C# strives to be a language that is explicit about program behavior and normally required the developer to write out what their intention is without ambiguities. This makes is a language that is easy for someone to read and understand once you learn it as there is little implicit behavior. I believe having this "gotcha" where keywords are only keywords if nothing else in scope is so-named makes the language harder to read and reason about in general.

There is also the reality that the design goals around C# Language versions and .NET Framework versions have changed. In the past it was paramount that developers could take a new language version update without updating the framework version they were targeting. With language features being increasingly tied to the runtime this makes less sense. We now strongly encourage developers to update both the language version and target framework together.

## Detailed design
[design]: #detailed-design

There is an existing concept in the language called "contextual keywords". I am not proposing doing away with this concept altogether just changing it so that a keywords "contextual-ness" is always able to be determined lexically. Take new (at the time of this writing) keyword `record`. We can still know if we are referring to the `record` keyword or some identifier named record based on the lexical context, there is no ambiguity. However `var`, according to the spec, requires us to check if there are types named var in scope:

[spec](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/statements#local-variable-declarations)
> In the context of a local variable declaration, the identifier var acts as a contextual keyword. When the local_variable_type is specified as var and no type named var is in scope,

Similarly `nameof` requires checking if a there are any identifiers called nameof in scope

[spec](https://docs.microsoft.com/dotnet/csharp/language-reference/language-specification/statements#local-variable-declarations)

> Because nameof is not a reserved keyword, a nameof expression is always syntactically ambiguous with an invocation of the simple name nameof. For compatibility reasons, if a name lookup of the name nameof succeeds, the expression is treated as an invocation_expression -- regardless of whether the invocation is legal. Otherwise it is a nameof_expression.

The implementation of this proposal would remove wording from the spec around name lookup collisions, and have a compliant compiler be able to fully determine keywords given only parsing information.

The following keywords would now error if developers attempted to use them as anything other than a keyword
- `var`
- `nameof`
- `dynamic`
- `_`

## Drawbacks
[drawbacks]: #drawbacks

This is a breaking change, if anyone was relying on this behavior in their code it would no longer compile. For most cases where a type named `var` or a method named `nameof` is in use the developer would need to change the usages to `@var` and `@nameof`.

## Alternatives
[alternatives]: #alternatives

We could opt to keep `_` as a contextual keyword that depends on name lookup rules as this is the change most likely to break real-world programs (see discussion on https://github.com/dotnet/csharplang/issues/1064)

**Comment by RikkiGibson** [comment]
I feel like it's worth mentioning that the "migration path" for most cases where a type named `var` or a method named `nameof` is in use is to change the usages to `@var` and `@nameof`.

**Reply by jmarolf** [reply]
ah, I'll add that to the `Drawbacks` section

**Comment by HaloFour** [comment]
Java:

![image](https://user-images.githubusercontent.com/10102290/108903358-3a407800-75eb-11eb-8a58-979900ec3a73.png)

**Comment by CyrusNajmabadi** [comment]
I believe that @jaredpar's position on this is that we could tie this to language version upgrades, similar to the breaking change we took for 'record' parsing.  If we do this, then we can take this codepaths out for later versions, and effectively limit support for previous versions.

**Reply by jaredpar** [reply]
Indeed that is my position. Now that customers actively have to opt into new language versions the value of `contextual keywords` has diminished greatly. Further they can serve to make the language harder to read and code. 

Take for example `_` which is either an identifier or a contextual keyword for a discard based on context. Our commitment to back compat here makes this feature harder to use because discards can't ... 

- Be used as method params, always an identifier there
- Be used in a lambda with a single arg cause that's an identifier, two args though both as `_` is a discard
- Be used in a lambda body where the dev used `_` as the single argument thinking it was a discard only to find it was an identifier. 

These rules exist because we were beholden to compatibility here and made it a contextual keyword. If on the other hand we made it a lexical keyword in C# 9 going forward we could have vastly simplified the implementation and rules: it's always a discard. 

That being said I want to acknowledge this is big step to take for the language and it has associated risks. The reason we made the first step with `record` is that we basically had no other choice due to the syntax we all wanted. Thus far that risk has panned out for us. Have to consider that perhaps `record` is special and we'd get into more trouble if we made say `_` a lexical keyword. I feel more confident about other contextual keywords like `var`, `dynamic`, `await`, etc ... but those have risks as well.

**Reply by HaloFour** [reply]
Does that mean that https://github.com/dotnet/csharplang/issues/1064 might be reconsidered?

**Reply by HaloFour** [reply]
Also Java:

![image](https://user-images.githubusercontent.com/10102290/108908129-f51f4480-75f0-11eb-8292-d2e932c15341.png)

**Reply by jaredpar** [reply]
> Does that mean that #1064 might be reconsidered?

If we took this approach then yes it would be reconsidered. 

Feel the need to re-emphasize that I'm definitely in favor of perusing some form of this but it's definitely not a consensus opinion of everyone in LDM. Best I can say is when I brought this up there wasn't a resounding "no" but more of a "huh, not the craziest thing jaredpar has said"

**Reply by HaloFour** [reply]
The funniest thing would be that if the C# compiler made it illegal to declare a type named `var` that the projects most likely to be broken are those that intentionally declare a type named `var` specifically to break `var` being used for local inference. 😁

**Reply by CyrusNajmabadi** [reply]
Yup. That sounds be the intent. Basically, this just isn't a philosophy we support for later Lang versions.

**Reply by jmarolf** [reply]
good point about `dynamic` and `_` being the other cases, are there others?

**Reply by CyrusNajmabadi** [reply]
Sorry, for dotnet format, i see no value in trying to format `nameof(...)` differently if it possibly didn't bind to the actual `nameof` operator.  It's just non-sensical to invest time and effort into that tool for pathological code like that.  If such code exists:

1. i attest they're a non-customer for such a tool.
2. that customer has options to disable us from formatting that section incorrectly.

`nameof` is effectively a c# keyword.  We absolutely made the call that we didn't want existing code to break if recompiled with a new compiler with that feature.  But we never made any such statement that we wouldn't want such code to be *reformatted* if they have that existing code construct.    

If you use `nameof` it's totally fine to expect that tools like formatters will see that and go: ok, that's clearly the `nameof operator` i'll format it appropriately.

**Reply by jmarolf** [reply]
you are right that nameof doesn't ever come into play, `var` on the other hand?

**Reply by CyrusNajmabadi** [reply]
well, we have no formatting rules that would be affected by 'var', so i'm not sure if it's relevant.

HOwever, say that 'var' *was* relevant for some formatting rule.  I woudl say it would 100% acceptable for our formatter to say:

"you wrote `var x`, so i'm doing to assume that that `var` is the type-inferred `var` and not some real type you defined in your system".

Our formatter doesn't have to support the pathological.  Specifically, people going out of their way to *subvert* the language.  If users do pathological stuff *and* want the formatter to support that, we're welcome to say: sorry, too bad, so sad :)   We would want to support the complex though.  So if you do have an array with 10,000 elements in it... well, that's extreme and complex, but not unreasonable.  So we should try to support *if* we can.  But people who abuse things like contextual keywords, with their own semantic types that conflict with them?  I do not think we have to care about that case at all.

**Reply by jmarolf** [reply]
> Our formatter doesn't have to support the pathological. 

right, my point is that _all_ formatters need to consider this today. If they just blindly use the APIs they for roslyn and expect to get keywords for this, they won't. It's by no means insurmountable, especially for us who have had years to tune things. If someone wants to color code on their blog what should they do if they want to handle contextual keywords? They can't just call roslyn apis and expect things to work. They need to learn about the contextual rules and decide what heuristics make sense for there use cases. I would prefer if "Is this a keyword" was a remarkably simple question to answer that required no education for users of roslyns apis.

Admittedly this desire only serves a small portion of the C# community and I think you could most definitely argue that this alone is not a strong enough argument to justify the change

**Reply by jmarolf** [reply]
If the roslyn apis had these heuristics built in, or had helpers that said "right but please honey, this is _definitely_ a keyword" then I think I wouldn't care. But this is something everyone writing tooling for C# needs to care about and think about today.

**Reply by CyrusNajmabadi** [reply]
> I would prefer if "Is this a keyword" was a remarkably simple question to answer that required no education for users of roslyns apis.

I'm 100% ok with that :)

**Comment by CyrusNajmabadi** [comment]
> The key motivator here is tooling. There are several tools such as dotnet-format which operate on syntax, but because of the way keywords work in the language today these tools need to get all the command-line arguments to the compiler to be accurate which typically means calling msbuild. Tooling authors are faced with an unfortunate choice of building slow but accurate tools or building fast but inaccurate ones.

I woudl be surprised if any formatting tooling actually cared about this (including for the nameof case).  In practice `nameof` will *100%* of the time be the C# language feature.  So it's totally fine for our tools to assume as such.  The same with 'var', 'dynamic' and the other semantic contextual keywords.

So i don't see this really buying much at the tooling level TBH, as tehse tools shoudl care about the practical case, not the pathological.

**Reply by jmarolf** [reply]
in Visual Studio _sure_ because we expect user interaction in some form or fashion. Tools like `dotnet-format` operate over millions of lines of code and people expect it to not just use heuristics to make decisions since there is no human interaction.

**Comment by jmarolf** [comment, marked as answer]
I've moves this proposal to an issue [here](https://github.com/dotnet/csharplang/issues/4460)
